internal/node/reverse_proxy/server.go

```diff
@@ -66,13 +70,40 @@ func (s *ProxyServer) Start(addr string, tlsConfig *tls.Config) (func(), error)
 
 	eg := &errgroup.Group{}
 	eg.Go(func() error {
-		var err error
+
+		// If TS_AUTHKEY is set, create a tsnet listener. Otherwise, create a normal tcp listener.
+		var listener net.Listener
+		if s.nodeConfig.TailscaleAuthkey() == "" {
+			ln, err := net.Listen("tcp", addr)
+			if err != nil {
+				return err
+			}
+			defer ln.Close()
+			listener = ln
+		} else {
+			tsnet := &tsnet.Server{
+				Hostname: constants.DefaultTSNetHostname,
+				Dir:      s.nodeConfig.TailScaleStatePath(),
+				Logf: func(msg string, args ...any) {
+					s.logger.Debug().Msgf(msg, args...)
+				},
+			}
+
+			defer tsnet.Close()
+
+			ln, err := tsnet.Listen("tcp", addr)
+			if err != nil {
+				return err
+			}
+			defer ln.Close()
+			listener = ln
+		}
+
 		if tlsConfig != nil {
-			err = httpServer.ListenAndServeTLS(s.nodeConfig.NodeCertPath(), s.nodeConfig.NodeKeyPath())
+			return httpServer.ServeTLS(listener, s.nodeConfig.NodeCertPath(), s.nodeConfig.NodeKeyPath())
```

Comment:
I don't quite understand the differences between using `Serve` vs `ListenAndServe`?

Reply:
Pretty much no difference except that you pass the listener in directly rather than letting the function create it. We need it in this case because we build a custom listener with `tsnet`